### PR TITLE
fix(plugin-coding-tools): bound web-fetch JSON extract path walk

### PR DESCRIPTION
**Pull request by byt61**
## Summary

- Fix `plugins/plugin-coding-tools/src/actions/web-fetch.ts` `WEB_FETCH` extract hint to bound its attacker-controlled dot-path walk the way `connector-json` (#23845) and `role-gate` (#23847) bound theirs. The previous `for (segment of path.split(".")) obj = obj[segment]` on attacker-controlled `path` could be driven with a 100k-segment or deeply-nested path to pin the model turn with unbounded path-walk work. `JSON.parse` produces ordinary JSON objects, so getters/Proxy traps are not reachable from the parsed response; the new walk is descriptor-only as defense-in-depth hardening.
- Bound walk: path length 1024, depth 16, segment 256, descriptor-only reflection (no getter/proxy), fail-safe `undefined` that falls back to full JSON so a fuzzy path never hard-fails. Preserves existing fallback-to-full-JSON semantics.
- Also bound the runtime fallback `packages/agent/src/runtime/actions/web-fetch.ts` which was still unbounded (`for (segment of path.split(".")) current=(current as Record)[segment]`) and registered via `eliza.ts:5210-5222` when coding-tools is not present. Now shares identical limits (1024/16/256, descriptor-only) and fails closed to full JSON. The fallback path preserves the complete guarded response body (`toWellFormedUnicode(body)`) — the guarded transport enforces the 262144-byte ceiling, so no 4k snippet truncation is added (that would revert #24134).
- Align `plugins/plugin-coding-tools/src/actions/web-fetch.test.ts` large-body wording to the actual `262144-byte safety ceiling` error (was stale `complete-capture safety limit`).
- Add deterministic boundary tests via the public action paths for both runtimes: valid nested path, missing path fallback, empty segment (`a..b`), >16 depth, >256 segment, >1024 path, and accessor fallback — all assert fallback to full JSON rather than hard-fail. `JSON.parse` produces plain data properties, so accessor/Proxy trap reachability is not gated by the descriptor check in these JSON tests (descriptor-only is defense-in-depth).

Same family as approved #23845/#23847 bounded walks — deterministic hosted-input hardening.

## Changes

| File | Change |
|------|--------|
| `plugins/plugin-coding-tools/src/actions/web-fetch.ts` | Bound extract path walk: length 1024, depth 16, segment 256, descriptor-only, fail-safe undefined |
| `packages/agent/src/runtime/actions/web-fetch.ts` | Mirror bound walk (1024/16/256, descriptor-only) and preserve complete body (`toWellFormedUnicode`) — no 4k truncation; update JSDoc to plain-JSON defense-in-depth and extract description to complete body |
| `plugins/plugin-coding-tools/src/actions/web-fetch.test.ts` | Align ceiling wording to `safety ceiling`; add 6 bound coverage tests |
| `packages/agent/src/runtime/actions/web-fetch.test.ts` | Add 7 bound coverage tests (including complete-body preservation and accessor fallback) |

## Verification

- Rebased onto `origin/develop@1b03eb54c0` — zero conflicts
- `bunx biome check --write` — 4 files clean
- `bunx tsc --noEmit --project packages/agent/tsconfig.json` — passed (only pre-existing `plugin-pty` any)
- `bunx tsc --noEmit --project plugins/plugin-coding-tools/tsconfig.json` — passed
- `bash -c 'cd /tmp/wt-24187 && ./node_modules/.bin/vitest run plugins/plugin-coding-tools/src/actions/web-fetch.test.ts'` — pre-bound 19 passed; post-bound adds 6 (25 total) — verified via main copy earlier: `bun x vitest run plugins/plugin-coding-tools/src/actions/web-fetch.test.ts` 25 passed after ceiling wording alignment
- `bash -c 'cd /tmp/wt-24187 && ./node_modules/.bin/vitest run packages/agent/src/runtime/actions/web-fetch.test.ts'` — 23 passed (including new 7 bounds with complete-body preservation)
- `git show HEAD:packages/agent/src/runtime/actions/web-fetch.ts` verifies bound constants, `toWellFormedUnicode(body)` complete body, and no `truncateWellFormed`/`_WEB_FETCH_SNIPPET_CHARS`
- `git show HEAD:plugins/plugin-coding-tools/src/actions/web-fetch.ts` verifies bound constants

## Evidence Gate

- [x] N/A - no UI surface; coding-tools web-fetch is headless extract logic.

- [x] N/A - same as before.

- [x] N/A - no user-facing flow.

- [x] Typecheck and vitest passed:

```log
bunx tsc --noEmit --project packages/agent/tsconfig.json
(no output — passed)

bun --bun x vitest run packages/agent/src/runtime/actions/web-fetch.test.ts
23 passed

bun --bun x vitest run plugins/plugin-coding-tools/src/actions/web-fetch.test.ts
25 passed
```

- [x] N/A - no frontend involved.

- [x] N/A - no prompt/model delegation change beyond extract hygiene.

- [x] Domain artifact = bound extract constants in both web-fetch paths + complete-body preservation, verified by git show

- [x] N/A - no rendered visual surface.

# Risks

Low — single extract path hardened in both registrations; fail-safe preserves prior fallback-to-full-JSON on invalid path, so fuzzy user hints never hard-fail. Complete-body preservation (no 4k snippet cap) restores #24134 guarantee — guarded transport is the ceiling.

# Background

# Documentation changes needed?

My changes do not require a change to the project documentation.

---

AI provider/model: openai / gpt-5.6-sol
Client / agent tooling: Codex Desktop
Contribution skill revision: elizaOS/eliza@1b03eb54c0:packages/skills/skills/contribute-to-eliza

### CODE REVIEW

**Review by lalalune — CHANGES_REQUESTED** [review]
Current-head review at `2f77f5e9b26070e47f13d49ea4d6c94dca7a075c`: the source change is small and the 1,024-character pre-check does bound the subsequent `split`, but this security/availability fix has no behavioral tests. Please add deterministic coverage for: a valid nested path; missing/accessor path preserving the documented full-JSON fallback; empty segments; >16 segments; >256-character segment; and >1,024-character path. The tests should exercise `WEB_FETCH`/`extractBody` through the public action path, not only duplicate the helper logic. Also narrow the PR wording: `JSON.parse` produces ordinary JSON objects, so getters/Proxy traps are not reachable from the parsed response; the demonstrated issue here is attacker-controlled path-walk work. Keep the PR ready for review while repairing.

File: plugins/plugin-coding-tools/src/actions/web-fetch.ts [hunk 1]
```diff
@@ -98,11 +98,32 @@ export function htmlToReadableText(html: string): string {
   return body;
 }
 
+const MAX_JSON_EXTRACT_DEPTH = 16;
+const MAX_JSON_EXTRACT_PATH_LENGTH = 1024;
+const MAX_JSON_EXTRACT_SEGMENT_LENGTH = 256;
+
 function resolveJsonPath(root: unknown, path: string): unknown {
+  if (path.length === 0 || path.length > MAX_JSON_EXTRACT_PATH_LENGTH)
```
Comment (standujar) [inline plugins/plugin-coding-tools/src/actions/web-fetch.ts:106]: [P1] Apply the bound to the runtime fallback too. This only hardens the coding-tools implementation: `packages/agent/src/runtime/actions/web-fetch.ts:80-87` still performs an unbounded `path.split(".")` plus bracket walk, and `packages/agent/src/runtime/eliza.ts:5210-5222` registers that action whenever coding-tools does not own `WEB_FETCH` (notably supported lean-chat/store-style profiles where the plugin is excluded). On this exact head, a public-action repro returned `42` for both a 17-segment extract and a 257-character segment, while this implementation correctly rejected those shapes and fell back to full JSON. The availability hardening therefore remains bypassed on a production registration path; share the bounded resolver or apply equivalent limits and public-action tests to both implementations.

AI provider/model: OpenAI / gpt-5.6-sol
Client / agent tooling: Codex desktop
Contribution skill revision: elizaOS/eliza@b687e3bbc9347c6d9d273b67c28a9b05dc52810b:packages/skills/skills/contribute-to-eliza
Attribution status: self-reported
— [standujar/json-extract]

**Review by standujar — CHANGES_REQUESTED** [review]
Current-head adversarial review at `2f77f5e9b26070e47f13d49ea4d6c94dca7a075c`.

One blocking finding is inline and is distinct from the existing request for missing boundary tests: the supported fallback `WEB_FETCH` still has the same unbounded extract walk.

Also fix the ownership metadata before merge: the body says `Fixes #23536`, but #23536 is the unrelated personal-assistant surrogate-pair issue already completed by #23537. GitHub currently exposes it as this PR's `closingIssuesReferences`; replace it with the actual owning issue (or open/link one for this scope).

Independent proof in a network-denied disposable sandbox:
- the new coding-tools path correctly fell back for a 17-segment path and a 257-character segment (2/2 focused adversarial checks);
- the registered agent fallback returned the extracted leaf `42` for both inputs instead of applying those bounds (2/2 parity checks failed as expected);
- Biome 2.5.5 reported the changed source clean;
- the existing coding-tools web-fetch suite reached 18/19, with its unrelated pre-existing response-limit wording assertion still expecting `complete-capture safety limit` while the source emits `262144-byte safety ceiling`.

AI provider/model: OpenAI / gpt-5.6-sol
Client / agent tooling: Codex desktop
Contribution skill revision: elizaOS/eliza@b687e3bbc9347c6d9d273b67c28a9b05dc52810b:packages/skills/skills/contribute-to-eliza
Attribution status: self-reported
— [standujar/json-extract]

**Review by lalalune — CHANGES_REQUESTED** [review]
Current-head blocker at `8af773c35dc82941eecabb631a470742e7b412d3`: the requested extract-path bounds and fallback-path parity are now implemented and tested, but this head also reintroduces a 4,000-character fallback body cap in `packages/agent/src/runtime/actions/web-fetch.ts`. That directly reverses the complete-context behavior merged in #24134 and now regression-tested by #24201; the guarded transport already enforces the explicit 262144-character full-or-error ceiling. Remove the `_WEB_FETCH_SNIPPET_CHARS` truncation change and its new 4k expectation while retaining the JSON-path bounds/tests. Also avoid claiming the accessor test exercises a getter: the public path JSON-parses the response into ordinary data properties, so that case only proves missing-path fallback.

**Review by lalalune — APPROVED** [review]
Independent exact-head review GREEN. Both production registrations bound the attacker-controlled path before split/walk (1024 code units, 16 segments, 256 per segment), use own data descriptors only, and preserve the established full-JSON fallback. JSON.parse is the sole data producer at both seams, so accessors/proxies are unreachable in ordinary production; arrays/numeric paths remain compatible. Submitted exact focused receipts are 25/25 plugin and 23/23 agent plus both package typechecks. I independently checked the exact four files with Biome and diff-check. Rebase to current develop was conflict-free and behavior blobs remained unchanged.
